src/Features/LanguageServer/Protocol/RequestDispatcher.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer
+{
+    /// <summary>
+    /// Aggregates handlers for the specified language and dispatches LSP requests
+    /// to the appropriate handler for the request.
+    /// </summary>
+    internal class RequestDispatcher
+    {
+        private readonly ImmutableDictionary<string, (IRequestHandler RequestHandler, ILspMethodMetadata Metadata)> _requestHandlers;
+
+        public RequestDispatcher(ImmutableArray<Lazy<AbstractRequestHandlerProvider, IRequestHandlerProviderMetadata>> requestHandlerProviders, string? languageName = null)
+        {
+            _requestHandlers = CreateMethodToHandlerMap(requestHandlerProviders.Where(rh => rh.Metadata.LanguageName == languageName));
+        }
+
+        private static ImmutableDictionary<string, (IRequestHandler, ILspMethodMetadata)> CreateMethodToHandlerMap(IEnumerable<Lazy<AbstractRequestHandlerProvider, IRequestHandlerProviderMetadata>> requestHandlerProviders)
+        {
+            var requestHandlerDictionary = ImmutableDictionary.CreateBuilder<string, (IRequestHandler, ILspMethodMetadata)>(StringComparer.OrdinalIgnoreCase);
+
+            // Create the actual request handlers from the providers.
+            var handlers = requestHandlerProviders.SelectMany(lazyProvider => lazyProvider.Value.CreateRequestHandlers());
+
+            // Store the request handlers in a dicitionary from method name (+ command name if workspace/executeCommand method) to handler instance.
+            foreach (var handler in handlers)
+            {
+                var lspMethodNameWithCommandName = handler.Metadata.MethodName + handler.Metadata.CommandName;
```

Comment:
Leaky, but also if we are concatenating it would be nicer in the logs for a `/` to go in the middle or something.